### PR TITLE
Make the lambda-variant of guardWith behave as a lazy initializer.

### PR DESCRIPTION
**Pull request by jasonwyatt**

### CODE REVIEW

File: java/arcs/core/util/Guard.kt [hunk 1]
```diff
@@ -70,15 +72,20 @@ class Guard<T>(private val mutex: Mutex, private val initialValue: T) {
  */
 private class GuardDelegate<T>(
     private val mutex: Mutex,
-    private var value: T
+    private val initialValue: () -> T
 ) : ReadWriteProperty<Any, T> {
+    private var valueHolder: ValueHolder<T>? = null
```
Comment (jblebrun) [inline java/arcs/core/util/Guard.kt]: Is the goal of this boxing to make it possible to set `null` on optional types while still preventing the initializer from running? (A test to verify setting `null` prevents initializer run might be nice to add too, in that case?)

**Comment by jblebrun** [comment]
Another test idea: verify that two successive gets without a set only result in one initializer call

File: java/arcs/core/util/Guard.kt (hunk 1 above)
Comment (jasonwyatt) [reply java/arcs/core/util/Guard.kt]: That's right

**Comment by jasonwyatt** [comment]
> Another test idea: verify that two successive gets without a set only result in one initializer call

That gets tested in the `initialValueLambda_isTreatedAsALazy_function` test scenario

File: java/arcs/core/util/Guard.kt (hunk 1 above)
Comment (jasonwyatt) [reply java/arcs/core/util/Guard.kt]: And great test idea - that caught a bug! Thanks!

**Pull request merged**